pkg/yurthub/proxy/remote/loadbalancer.go

```diff
@@ -163,10 +184,170 @@ func (lb *loadBalancer) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 		http.Error(rw, "could not pick one healthy backends, try again to go through local proxy.", http.StatusInternalServerError)
 		return
 	}
-	klog.V(3).Infof("picked backend %s by %s for request %s", rp.Name(), lb.algo.Name(), util.ReqString(req))
+	klog.V(3).Infof("picked backend %s by %s for request %s", rp.Name(), lb.algo.Name(), hubutil.ReqString(req))
 	rp.ServeHTTP(rw, req)
 }
+
+func (lb *loadBalancer) errorHandler(rw http.ResponseWriter, req *http.Request, err error) {
+	klog.Errorf("remote proxy error handler: %s, %v", hubutil.ReqString(req), err)
+	if lb.localCacheMgr == nil || !lb.localCacheMgr.CanCacheFor(req) {
+		rw.WriteHeader(http.StatusBadGateway)
+		return
+	}
+
+	ctx := req.Context()
+	if info, ok := apirequest.RequestInfoFrom(ctx); ok {
+		if info.Verb == "get" || info.Verb == "list" {
+			if obj, err := lb.localCacheMgr.QueryCache(req); err == nil {
+				hubutil.WriteObject(http.StatusOK, obj, rw, req)
+				return
+			}
+		}
+	}
+	rw.WriteHeader(http.StatusBadGateway)
+}
+
+func (lb *loadBalancer) modifyResponse(resp *http.Response) error {
+	if resp == nil || resp.Request == nil {
+		klog.Infof("no request info in response, skip cache response")
+		return nil
+	}
+
+	req := resp.Request
+	ctx := req.Context()
+
+	// re-added transfer-encoding=chunked response header for watch request
+	info, exists := apirequest.RequestInfoFrom(ctx)
+	if exists {
+		if info.Verb == "watch" {
+			klog.V(5).Infof("add transfer-encoding=chunked header into response for req %s", hubutil.ReqString(req))
+			h := resp.Header
+			if hv := h.Get("Transfer-Encoding"); hv == "" {
+				h.Add("Transfer-Encoding", "chunked")
+			}
+		}
+	}
+
+	if resp.StatusCode >= http.StatusOK && resp.StatusCode <= http.StatusPartialContent {
+		// prepare response content type
+		reqContentType, _ := hubutil.ReqContentTypeFrom(ctx)
+		respContentType := resp.Header.Get("Content-Type")
+		if len(respContentType) == 0 {
+			respContentType = reqContentType
+		}
+		ctx = hubutil.WithRespContentType(ctx, respContentType)
+		req = req.WithContext(ctx)
+
+		// filter response data
+		if lb.filterManager != nil {
+			if ok, runner := lb.filterManager.FindRunner(req); ok {
+				wrapBody, needUncompressed := hubutil.NewGZipReaderCloser(resp.Header, resp.Body, req, "filter")
+				size, filterRc, err := runner.Filter(req, wrapBody, lb.stopCh)
+				if err != nil {
+					klog.Errorf("failed to filter response for %s, %v", hubutil.ReqString(req), err)
+					return err
+				}
+				resp.Body = filterRc
+				if size > 0 {
+					resp.ContentLength = int64(size)
+					resp.Header.Set("Content-Length", fmt.Sprint(size))
+				}
+
+				// after gunzip in filter, the header content encoding should be removed.
+				// because there's no need to gunzip response.body again.
+				if needUncompressed {
+					resp.Header.Del("Content-Encoding")
+				}
+			}
+		}
+
+		if lb.workingMode == hubutil.WorkingModeEdge {
+			// cache resp with storage interface
+			lb.cacheResponse(req)
```

Comment:
`hubutil.NewGZipReaderCloser` should be considered before cacheResponse.

Reply:
fixed